### ISSUE TITLE
"Number of open connections" does not go down after Timeout

### ISSUE THREAD

**Issue by kuk** [tinyproxy/tinyproxy#414]
### Issue

Threads do not terminate after Timeout seconds.

Using tinyproxy 1.11.0.

### How to reproduce

tinyproxy.conf:
```
Port 9482

Timeout 10
```

app.py web application that for every request sleeps for 2 minutes, listens on 8080:
```
from asyncio import sleep

from aiohttp import web


async def handler(request):
    await sleep(120)
    return web.Response(text='ok')


app = web.Application()
app.add_routes([
    web.get('/', handler)
])
web.run_app(app)
```

Issue request to localhost:8080 thought localhost:9482 with 5 seconds timeout. Fail because app responds for 2 minutes:
```
curl http://localhost:8080 --proxy localhost:9482 --max-time 5
curl: (28) Operation timed out after 5002 milliseconds with 0 bytes received
```

Repeat 10 times. Wait 10 seconds (tinyproxy.conf Timeout). Request tinyproxy.stats, expect 0 open connections, but get 10 + 1 for stats request

```
curl http://tinyproxy.stats --proxy localhost:9482

Number of open connections: 11<br />
Number of requests: 11<br />
Number of bad connections: 0<br />
Number of denied connections: 0<br />
Number of refused connections due to high load: 0
```

Terminate app.py. Repeat tinyproxy.stats request:
```
Number of open connections: 1<br />
Number of requests: 12<br />
Number of bad connections: 10<br />
Number of denied connections: 0<br />
Number of refused connections due to high load: 0
```

### Question

Is it correct behaviour? Should tinyproxy wait for response from site even after Timeout? Even if original request timedout?

After 100 such request tinyproxy just stops responding

**Comment by rofl0r** [comment]
should not happen. can you start tinyproxy via `strace -f`, make a single such connection and paste output from where it gets stuck ?

**Comment by kuk** [comment]
After I kill app.py it continues

```
[pid    61] recvfrom(7, "", 512, MSG_PEEK, NULL, NULL) = 0
[pid    61] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705381, tv_nsec=705973100}) = 0
[pid    61] getpid()                    = 60
[pid    61] write(5, "INFO      Jan 20 19:03:01.705 [6"..., 129) = 129
[pid    61] fsync(5)                    = 0
[pid    61] close(6)                    = 0
[pid    61] close(7)                    = 0
[pid    61] madvise(0x7f684a014000, 241664, MADV_DONTNEED) = 0
[pid    61] exit(0)                     = ?
[pid    61] +++ exited with 0 +++

```

Relaunch app.py, send another request
```

<... poll resumed>)                     = 1 ([{fd=3, revents=POLLIN}])
accept(3, {sa_family=AF_INET, sin_port=htons(61852), sin_addr=inet_addr("172.17.0.1")}, [16]) = 6
mmap(NULL, 266240, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f6849fae000
mprotect(0x7f6849faf000, 262144, PROT_READ|PROT_WRITE) = 0
clone(child_stack=0x7f6849fedfb0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[62]
, tls=0x7f6849fee700, child_tidptr=0x7f6849fee9d0) = 62
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, -1strace: Process 62 attached
 <unfinished ...>
[pid    62] set_robust_list(0x7f6849fee9e0, 24) = 0
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=599253400}) = 0
[pid    62] getpid()                    = 60
[pid    62] write(5, "CONNECT   Jan 20 19:04:48.599 [6"..., 76) = 76
[pid    62] fsync(5)                    = 0
[pid    62] setsockopt(6, SOL_SOCKET, SO_SNDTIMEO_OLD, "\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
[pid    62] setsockopt(6, SOL_SOCKET, SO_RCVTIMEO_OLD, "\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
[pid    62] recvfrom(6, "GET http://host.docker.internal:"..., 512, MSG_PEEK, NULL, NULL) = 208
[pid    62] recvfrom(6, "GET http://host.docker.internal:"..., 48, 0, NULL, NULL) = 48
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=608988100}) = 0
[pid    62] getpid()                    = 60
[pid    62] write(5, "CONNECT   Jan 20 19:04:48.608 [6"..., 112) = 112
[pid    62] fsync(5)                    = 0
[pid    62] recvfrom(6, "Host: host.docker.internal:8080\r"..., 512, MSG_PEEK, NULL, NULL) = 160
[pid    62] recvfrom(6, "Host: host.docker.internal:8080\r"..., 33, 0, NULL, NULL) = 33
[pid    62] recvfrom(6, "Proxy-Authorization: Basic cHJve"..., 512, MSG_PEEK, NULL, NULL) = 127
[pid    62] recvfrom(6, "Proxy-Authorization: Basic cHJve"..., 57, 0, NULL, NULL) = 57
[pid    62] recvfrom(6, "User-Agent: curl/7.64.1\r\nAccept:"..., 512, MSG_PEEK, NULL, NULL) = 70
[pid    62] recvfrom(6, "User-Agent: curl/7.64.1\r\n", 25, 0, NULL, NULL) = 25
[pid    62] recvfrom(6, "Accept: */*\r\nProxy-Connection: K"..., 512, MSG_PEEK, NULL, NULL) = 45
[pid    62] recvfrom(6, "Accept: */*\r\n", 13, 0, NULL, NULL) = 13
[pid    62] recvfrom(6, "Proxy-Connection: Keep-Alive\r\n\r\n", 512, MSG_PEEK, NULL, NULL) = 32
[pid    62] recvfrom(6, "Proxy-Connection: Keep-Alive\r\n", 30, 0, NULL, NULL) = 30
[pid    62] recvfrom(6, "\r\n", 512, MSG_PEEK, NULL, NULL) = 2
[pid    62] recvfrom(6, "\r\n", 2, 0, NULL, NULL) = 2
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=618788800}) = 0
[pid    62] getpid()                    = 60
[pid    62] write(5, "INFO      Jan 20 19:04:48.618 [6"..., 79) = 79
[pid    62] fsync(5)                    = 0
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=624663300}) = 0
[pid    62] getpid()                    = 60
[pid    62] write(5, "INFO      Jan 20 19:04:48.624 [6"..., 94) = 94
[pid    62] fsync(5)                    = 0
[pid    62] stat("/etc/resolv.conf", {st_mode=S_IFREG|0644, st_size=97, ...}) = 0
[pid    62] openat(AT_FDCWD, "/etc/hosts", O_RDONLY|O_CLOEXEC) = 7
[pid    62] lseek(7, 0, SEEK_CUR)       = 0
[pid    62] fstat(7, {st_mode=S_IFREG|0644, st_size=174, ...}) = 0
[pid    62] read(7, "127.0.0.1\tlocalhost\n::1\tlocalhos"..., 4096) = 174
[pid    62] lseek(7, 0, SEEK_CUR)       = 174
[pid    62] read(7, "", 4096)           = 0
[pid    62] close(7)                    = 0
[pid    62] clock_gettime(CLOCK_MONOTONIC, {tv_sec=424853, tv_nsec=780461804}) = 0
[pid    62] clock_gettime(CLOCK_MONOTONIC, {tv_sec=424853, tv_nsec=782153904}) = 0
[pid    62] socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 7
[pid    62] setsockopt(7, SOL_IP, IP_RECVERR, [1], 4) = 0
[pid    62] connect(7, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.65.5")}, 16) = 0
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=653303900}) = 0
[pid    62] poll([{fd=7, events=POLLOUT}], 1, 0) = 1 ([{fd=7, revents=POLLOUT}])
[pid    62] sendmmsg(7, [{msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\344\37\1\0\0\1\0\0\0\0\0\0\4host\6docker\10interna"..., iov_len=38}], msg_iovlen=1, msg_control
len=0, msg_flags=0}, msg_len=38}, {msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\346[\1\0\0\1\0\0\0\0\0\0\4host\6docker\10interna"..., iov_len=38}], msg_iovlen=1, msg_
controllen=0, msg_flags=0}, msg_len=38}], 2, MSG_NOSIGNAL) = 2
[pid    62] poll([{fd=7, events=POLLIN}], 1, 5000) = 1 ([{fd=7, revents=POLLIN}])
[pid    62] ioctl(7, FIONREAD, [38])    = 0
[pid    62] recvfrom(7, "\346[\201\200\0\1\0\0\0\0\0\0\4host\6docker\10interna"..., 2048, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.65.5")}, [28->16]) =
 38
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=669192500}) = 0
[pid    62] poll([{fd=7, events=POLLIN}], 1, 4984) = 1 ([{fd=7, revents=POLLIN}])
[pid    62] ioctl(7, FIONREAD, [54])    = 0
[pid    62] recvfrom(7, "\344\37\201\200\0\1\0\1\0\0\0\0\4host\6docker\10interna"..., 65536, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.65.5")}, [28->16]
) = 54
[pid    62] close(7)                    = 0
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=676364700}) = 0
[pid    62] getpid()                    = 60
[pid    62] write(5, "INFO      Jan 20 19:04:48.676 [6"..., 97) = 97
[pid    62] fsync(5)                    = 0
[pid    62] socket(AF_INET, SOCK_STREAM, IPPROTO_TCP) = 7
[pid    62] connect(7, {sa_family=AF_INET, sin_port=htons(8080), sin_addr=inet_addr("192.168.65.2")}, 16) = 0
[pid    62] clock_gettime(CLOCK_REALTIME, {tv_sec=1642705488, tv_nsec=688496000}) = 0
[pid    62] getpid()                    = 60
[pid    62] write(5, "CONNECT   Jan 20 19:04:48.688 [6"..., 115) = 115
[pid    62] fsync(5)                    = 0
[pid    62] sendto(7, "GET / HTTP/1.0\r\nHost: host.docke"..., 68, MSG_NOSIGNAL, NULL, 0) = 68
[pid    62] sendto(7, "User-Agent: curl/7.64.1\r\n", 25, MSG_NOSIGNAL, NULL, 0) = 25
[pid    62] sendto(7, "Accept: */*\r\n", 13, MSG_NOSIGNAL, NULL, 0) = 13
[pid    62] sendto(7, "\r\n", 2, MSG_NOSIGNAL, NULL, 0) = 2
[pid    62] recvfrom(7,
```

**Comment by rofl0r** [comment]
thanks, very conclusive. could try out whether my PR https://github.com/tinyproxy/tinyproxy/pull/415 fixes the issue ?

**Comment by kuk** [comment]
Yes, it fixes, thank you!

**Comment by rofl0r** [comment]
thanks, merged #415

**Comment by kuk** [comment]
May also be the fix for https://github.com/tinyproxy/tinyproxy/issues/274
